source/code/plugin/out_mdm.rb

```diff
@@ -0,0 +1,219 @@
+      ApplicationInsightsUtility.sendCustomEvent("AKSCustomMetricsMDMPluginStart", {})
+    end
+
+    # get the access token only if the time to expiry is less than 5 minutes
+    def get_access_token
+      if @cached_access_token.to_s.empty? || (Time.now + 5*60 > @token_expiry_time) # token is valid for 60 minutes. Refresh token 5 minutes from expiration
+        @log.info "Refreshing access token for out_mdm plugin.."
+        token_uri = URI.parse(@token_url)
+        http_access_token = Net::HTTP.new(token_uri.host, token_uri.port)
+        http_access_token.use_ssl = true
+        token_request = Net::HTTP::Post.new(token_uri.request_uri)
+        token_request.set_form_data(
+          {
+            'grant_type' => @@grant_type, 
+            'client_id' => @data_hash['aadClientId'], 
+            'client_secret' => @data_hash['aadClientSecret'],
+            'resource' => @@token_resource_url
+            }
+        )
+        
+        token_response = http_access_token.request(token_request)
+        # Handle the case where the response is not 200 
+        parsed_json = JSON.parse(token_response.body)
+        @token_expiry_time = Time.now + 59*60 # set the expiry time to be ~one hour from current time
+        @cached_access_token = parsed_json['access_token']
+      end
+      @cached_access_token
+    end 
+
+    def write_status_file(success, message)
+      fn = '/var/opt/microsoft/omsagent/log/MDMIngestion.status'
+      status = '{ "operation": "MDMIngestion", "success": "%s", "message": "%s" }' % [success, message]
+      begin
+        File.open(fn,'w') { |file| file.write(status) }
+      rescue => e
+        @log.debug "Error:'#{e}'"
+        ApplicationInsightsUtility.sendExceptionTelemetry(e.backtrace)
+      end
+    end
+
+    # This method is called when an event reaches to Fluentd.
+    # Convert the event to a raw string.
+    def format(tag, time, record)
+      if record != {}
+        @log.trace "Buffering #{tag}"
+        return [tag, record].to_msgpack
+      else
+        return ""
+      end
+    end
+
+    # This method is called every flush interval. Send the buffer chunk to MDM. 
+    # 'chunk' is a buffer chunk that includes multiple formatted records
+    def write(chunk)
+      if !@first_post_attempt_made || (Time.now > @last_post_attempt_time + retry_mdm_post_wait_minutes*60)
+        post_body = []
+        chunk.msgpack_each {|(tag, record)|
+          post_body.push(record.to_json)
+        }
+        send_to_mdm post_body
+      else
+        @log.info "Last Failed POST attempt to MDM was made #{((Time.now - @last_post_attempt_time)/60).round(1)} min ago. This is less than the current retry threshold of #{@retry_mdm_post_wait_minutes} min. NO-OP"
+      end
+    end
+
+    def send_to_mdm(post_body) 
+      begin
+        access_token = get_access_token
+        uri = URI.parse(@@post_request_url)
+        http = Net::HTTP.new(uri.host, uri.port)
```

Comment:
Make httpclient static and reuse so you dont have to new up every time.

Reply:
Made it an instance variable.